### PR TITLE
Data utils

### PR DESCRIPTION
**Pull request by teresamg**

### CODE REVIEW

**Comment by teresamg** [comment]
Hey @oesteban, do you know what this error is? 

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [44], in <cell line: 1>()
----> 1 transformed_nii = (~xfms).apply(nii)
File ~/miniconda3/envs/eddymotion/lib/python3.8/site-packages/nitransforms/linear.py:440, in LinearTransformsMapping.apply(self, spatialimage, reference, order, mode, cval, prefilter, output_dtype)
    437 data = np.squeeze(np.asanyarray(spatialimage.dataobj))
    438 output_dtype = output_dtype or data.dtype
--> 440 ycoords = self.map(_ref.ndcoords.T)
    441 targets = ImageGrid(spatialimage).index(  # data should be an image
    442     _as_homogeneous(np.vstack(ycoords), dim=_ref.ndim)
    443 )
    445 if data.ndim == 4:
AttributeError: 'NoneType' object has no attribute 'ndcoords'

**Comment by oesteban** [comment]
Yup, you need to set a reference image for the transform

**Comment by teresamg** [comment]
@oesteban The tests seem to be passing now, any thoughts or is this ready to merge?

**Comment by arokem** [comment]
Looks like the test failure is unrelated to this PR. Just one of those intermittent/random failures of the registration.

File: src/eddymotion/data/utils.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    nii_t_img : :obj:`Nifti1Image`
+        Transformed Nifti1Image data
+
+    """
+    transformed_nii = np.zeros(nii.shape)
+    for ii, bvecnii in enumerate(nib.four_to_three(nii)):
+        xfms = nt.linear.Affine(em_affines[ii])
+        transformed_nii[..., ii] = (~xfms).apply(bvecnii, reference=nii).get_fdata()
+
+    nii_t_img = nib.Nifti1Image(transformed_nii, nii.affine)
```
Comment (oesteban) [inline src/eddymotion/data/utils.py:32]: Let nibabel figure out format and metadata
```suggestion
    nii_t_img = nii.__class__(transformed_nii, nii.affine, nii.header)
```

File: src/eddymotion/data/utils.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    output_filename : :obj:`str`, optional
+        String specifying filepath to which to save transformed Nifti1Image data
+
+    Returns
+    -------
+    nii_t_img : :obj:`Nifti1Image`
+        Transformed Nifti1Image data
+
+    """
+    transformed_nii = np.zeros(nii.shape)
```
Comment (oesteban) [inline src/eddymotion/data/utils.py:27]: Let's not override the internal datatype of the original image.
```suggestion
    transformed_nii = np.zeros_like(np.asanyarray(nii.dataobj))

```

File: src/eddymotion/data/utils.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    for ii, bvecnii in enumerate(nib.four_to_three(nii)):
+        xfms = nt.linear.Affine(em_affines[ii])
+        transformed_nii[..., ii] = (~xfms).apply(bvecnii, reference=nii).get_fdata()
```
Comment (oesteban) [inline src/eddymotion/data/utils.py:30]: Like before, let nibabel handle the datatype
```suggestion
        transformed_nii[..., ii] = np.asanyarray(
            (~xfms).apply(bvecnii, reference=nii).dataobj
        )

```

File: src/eddymotion/data/utils.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    nii_t_img = nib.Nifti1Image(transformed_nii, nii.affine)
+
+    if output_filename is not None:
+        # Ensure directories in output_filename exist
+        Path(output_filename).parent.mkdir(exist_ok=True)
+
+        # Save as .nii
+        nib.save(nii_t_img, output_filename)
```
Comment (oesteban) [inline src/eddymotion/data/utils.py:39]:
```suggestion
        nii_t_img.to_filename(output_filename)
```

**Review by oesteban — APPROVED** [review]

File: src/eddymotion/data/utils.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    nii_t_img : :obj:`Nifti1Image`
+        Transformed Nifti1Image data
+
+    """
+    transformed_nii = np.zeros_like(np.asanyarray(nii.dataobj))
+
+    for ii, bvecnii in enumerate(nib.four_to_three(nii)):
+        xfms = nt.linear.Affine(em_affines[ii])
+        transformed_nii[..., ii] = np.asanyarray(
+            (~xfms).apply(bvecnii, reference=nii).dataobj
```
Comment (arokem) [inline src/eddymotion/data/utils.py:32]: I think that the CI is complaining about how many blank lines there are here:

```suggestion
            (~xfms).apply(bvecnii, reference=nii).dataobj
```

File: src/eddymotion/data/utils.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        )
```
Comment (arokem) [inline src/eddymotion/data/utils.py:33]: I think it doesn't like how many blank lines there are here:

```suggestion
        )
```

Comment (arokem) [reply src/eddymotion/data/utils.py:33]: That's what this comment is about, I believe: https://github.com/nipreps/eddymotion/actions/runs/4308645620/jobs/7515142254#step:5:8

**Pull request merged**